### PR TITLE
Implement `chcpu`

### PR DESCRIPTION
**Pull request by alxndrv**
This PR implements pretty much the entirety of `chcpu`.

Some things still missing:
- Tests: I'm not entirely sure how to properly test something which is so tightly coupled to system internals.
- Exit codes: `chcpu` should return a non-zero exit code in some edge cases. This implementation does check for most such cases, but pretty much always returns a success exit-code. I'm guessing that `uucore::error::UResult` can be used to solve this somehow without having to manually call `std::process::exit()`.

### CODE REVIEW

**Comment by sylvestre** [comment]
in general, to test such things, you either try to mock up
and you do your best to look at the error management

File: src/uu/chcpu/src/chcpu.rs [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+            }
+        } else {
+            let idx = part.parse::<usize>().expect("Invalid CPU index value");
+            out.push(idx);
+        }
+    }
+    out
+}
+
+#[test]
```
Comment (sylvestre) [inline src/uu/chcpu/src/chcpu.rs:40]: could you please move the tests at the end of the file? thanks

File: src/uu/chcpu/src/chcpu.rs [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+        true
+    }
+
+    // CPUs which are not hot-pluggable will not have the `/online` file in their directory
+    fn is_hotpluggable(&self) -> bool {
+        self.get_path().join("online").exists()
+    }
+
+    fn is_online(&self) -> bool {
+        if let Ok(state) = fs::read_to_string(self.get_path().join("online")) {
```
Comment (sylvestre) [inline src/uu/chcpu/src/chcpu.rs:77]: I think it can be simplified like

    fs::read_to_string(self.get_path().join("online"))
        .map(|state| match state.trim() {
            "0" => false,
            "1" => true,
            other => panic!("Unrecognized CPU online state: {}", other),
        })
        .unwrap_or(true) 

no ?

File: src/uu/chcpu/src/chcpu.rs [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+            return;
+        }
+
+        if self.is_online() {
+            println!("CPU {} is already enabled", self.0);
+            return;
+        }
+
+        let result =
+            File::create(self.get_path().join("online")).and_then(|mut f| f.write_all(b"1"));
```
Comment (sylvestre) [inline src/uu/chcpu/src/chcpu.rs:103]: looks like you are duplicating the code a bit in these 3 functions.
what about creating something like ?

fn write_cpu_state(path: PathBuf, value: &[u8], action: &str, cpu_id: u32) {
    let result = File::create(path).and_then(|mut f| f.write_all(value));
    match result {
        Ok(_) => println!("CPU {} {}", cpu_id, action),
        Err(e) => println!("CPU {} {} failed: {}", cpu_id, action, e.kind()),
    }
}

File: src/uu/chcpu/src/chcpu.rs [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+    fn enable(&self) {
+        if !self.is_hotpluggable() {
+            println!("CPU {} is not hot-pluggable", self.0);
+            return;
+        }
+
+        if self.is_online() {
+            println!("CPU {} is already enabled", self.0);
+            return;
+        }
```
Comment (sylvestre) [inline src/uu/chcpu/src/chcpu.rs:100]: you can also dedup this

File: src/uu/chcpu/src/chcpu.rs [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+}
+
+fn deconfigure_cpus(cpu_list: &str) {
+    let to_deconfigure = parse_cpu_list(cpu_list).into_iter().map(Cpu);
+    for cpu in to_deconfigure {
+        if cpu.ensure_exists() {
+            cpu.deconfigure();
+        }
+    }
+}
```
Comment (sylvestre) [inline src/uu/chcpu/src/chcpu.rs:245]: what about something like ?

```suggestion
fn process_cpus<F>(cpu_list: &str, action: F)
where
    F: Fn(&Cpu)
{
    parse_cpu_list(cpu_list)
        .into_iter()
        .map(Cpu)
        .filter(|cpu| cpu.ensure_exists())
        .for_each(|cpu| action(&cpu));
}

// Usage becomes:
fn enable_cpus(cpu_list: &str) {
    process_cpus(cpu_list, |cpu| cpu.enable());
}

fn disable_cpus(cpu_list: &str) {
    process_cpus(cpu_list, |cpu| cpu.disable());
}

fn configure_cpus(cpu_list: &str) {
    process_cpus(cpu_list, |cpu| cpu.configure());
}

fn deconfigure_cpus(cpu_list: &str) {
    process_cpus(cpu_list, |cpu| cpu.deconfigure());
}
```

**Comment by alxndrv** [comment]
@sylvestre Thanks for the review, I'll implement your suggestions a bit later. Since we're on the topic of deduping: there's some code here that's copy-pasted from my work on `lscpu` (the `parse_cpu_list()` function). Should there be some kind of "lib" crate for common stuff like this? These kinds of CPU lists come up in a bunch of Linux utils and theres no sense in implementing it all over again every time. There's other similar examples: for example the mount-related utils all have to parse mount entries from `/proc/<pid>/mountinfo` and could benefit from having a shared lib.

File: src/uu/chcpu/src/chcpu.rs (hunk 4 above)
Comment (alxndrv) [reply src/uu/chcpu/src/chcpu.rs:100]: Don't think this one can really be deduped? The second check is the inverse of the one here in the case of `Cpu::disable()`, (and the messages are different too).

**Comment by sylvestre** [comment]
> Should there be some kind of "lib" crate for common stuff like this? 

yes, we try to publish content in uucore
https://docs.rs/uucore/latest/uucore/

**Comment by alxndrv** [comment]
Updated the implementation to make use of `UResult`s in order to return correct non-zero exit codes in most failure scenarios. Also added some basic test scenarios and cleaned up the code a bit.

**Comment by alxndrv** [comment]
> > Should there be some kind of "lib" crate for common stuff like this?
> 
> yes, we try to publish content in uucore https://docs.rs/uucore/latest/uucore/

In regards to this, good to know. I can look into adding the shared CPU list parsing functionality to that later, and then replace the implementations present in `util-linux` in a separate PR. Does that sound fine?

**Comment by sylvestre** [comment]
I am really really sorry but I took this https://github.com/uutils/util-linux/pull/234 instead. it was more complete :(

sorry again :(
